subprojects/core/src/main/java/org/gradle/internal/execution/fingerprint/impl/DefaultFileCollectionFingerprinterRegistry.java

```diff
@@ -17,29 +17,61 @@
 package org.gradle.internal.execution.fingerprint.impl;
 
 import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
 import com.google.common.collect.Maps;
+import org.gradle.api.tasks.ClasspathNormalizer;
+import org.gradle.api.tasks.CompileClasspathNormalizer;
+import org.gradle.api.tasks.FileNormalizer;
 import org.gradle.internal.execution.fingerprint.FileCollectionFingerprinter;
 import org.gradle.internal.execution.fingerprint.FileCollectionFingerprinterRegistry;
 import org.gradle.internal.execution.fingerprint.FileNormalizationSpec;
+import org.gradle.internal.fingerprint.DirectorySensitivity;
+import org.gradle.internal.fingerprint.IgnoredPathInputNormalizer;
+import org.gradle.internal.fingerprint.LineEndingNormalization;
 
 import java.util.Collection;
 import java.util.Map;
+import java.util.Set;
 
 import static org.gradle.internal.execution.fingerprint.impl.DefaultFileNormalizationSpec.from;
 
 public class DefaultFileCollectionFingerprinterRegistry implements FileCollectionFingerprinterRegistry {
     private final Map<FileNormalizationSpec, FileCollectionFingerprinter> fingerprinters;
 
+    /**
+     * These are the normalizers where it does not make sense to ignore empty directories.  If the user specifies non-sensical empty directory
+     * normalization (e.g. via the runtime api) we will ignore that criteria when selecting a fingerprinter.
+     */
+    private final static Set<Class<? extends FileNormalizer>> DIRECTORY_INSENSITIVE_NORMALIZERS = ImmutableSet.of(
+        ClasspathNormalizer.class,
+        CompileClasspathNormalizer.class,
+        IgnoredPathInputNormalizer.class
+    );
```

Comment:
It feels like we should warn the user in some way. Since this annotation is already released in a previous version, we should probably deprecate the conflicting usages.

Reply:
See my comments above.